rpc/Transceiver.java

```diff
@@ -160,49 +122,137 @@ public boolean isOpen() {
         return !responseListener.terminated.get();
     }
 
+    /**
+     * Interface for collecting responses from a specific request
+     */
+    interface ResponseCollector {
+        /**
+         * Collect a response.
+         * @param response the next response for this collector to collect.
+         * @return true if this is the last response, false if more responses are expected.
+         */
+        boolean onResponse(Response response);
+    }
+
+    /**
+     * Simple response collector for when a single result is expected.
+     */
+    private static class SingleResponseCollector implements ResponseCollector {
+        private volatile Response response;
+        private final CountDownLatch latch = new CountDownLatch(1);
```

Comment:
why do we need this? Is it for corectness?

Reply:
It's because the collector receives messages from a pushing thread and returns them to pulling thread, `CountDownLatch` is used to block the pulling thread until the pushing thread has delivered the value.